File: 757. pyramid-transition-matrix.java
import java.util.*;

class Solution {
    private Map<String, List<Character>> mp;
    private Map<String, Boolean> memo;
    
    public boolean pyramidTransition(String bottom, List<String> allowed) {
        // Build a mapping: pair -> [possible_top_blocks]
        mp = new HashMap<>();
        for (String a : allowed) {
            String key = a.substring(0, 2);
            mp.computeIfAbsent(key, k -> new ArrayList<>()).add(a.charAt(2));
        }
        
        memo = new HashMap<>();
        return dfs(bottom);
    }
    
    private boolean dfs(String row) {
        // If we've reached a single block, pyramid is complete
        if (row.length() == 1) {
            return true;
        }
        
        // Check if we've already computed this row
        if (memo.containsKey(row)) {
            return memo.get(row);
        }
        
        // Try to build the next level
        boolean result = build(row, 0, "");
        memo.put(row, result);
        return result;
    }
    
    private boolean build(String row, int i, String path) {
        // If we've finished building the next level
        if (i == row.length() - 1) {
            return dfs(path);
        }
        
        // Get the pair of blocks at position i and i+1
        String pair = row.substring(i, i + 2);
        
        // Check if this pair has any allowed blocks
        if (!mp.containsKey(pair)) {
            return false;
        }
        
        // Try all possible blocks that can be placed on top
        for (char ch : mp.get(pair)) {
            if (build(row, i + 1, path + ch)) {
                return true;
            }
        }
        
        return false;
    }
    
    public static void main(String[] args) {
        Solution solution = new Solution();
        
        String bottom1 = "BCD";
        List<String> allowed1 = Arrays.asList("BCC", "CDE", "CEA", "FFF");
        boolean result1 = solution.pyramidTransition(bottom1, allowed1);
        System.out.println("Test 1: bottom='" + bottom1 + "', allowed=" + allowed1);
        System.out.println("Output: " + result1 + ", Expected: true\n");
    }
}

File: 757. pyramid-transition-matrix.py
from typing import List
from collections import defaultdict

class Solution:
    def pyramidTransition(self, bottom: str, allowed: List[str]) -> bool:
        mp = defaultdict(list)
        for a in allowed:
            mp[a[:2]].append(a[2])

        memo = {}

        def dfs(row: str) -> bool:
            if len(row) == 1:
                return True
            if row in memo:
                return memo[row]

            def build(i: int, path: str) -> bool:
                if i == len(row) - 1:
                    return dfs(path)
                pair = row[i:i+2]
                if pair not in mp:
                    return False
                for ch in mp[pair]:
                    if build(i + 1, path + ch):
                        return True
                return False

            memo[row] = build(0, "")
            return memo[row]

        return dfs(bottom)


# Test case
solution = Solution()

bottom1 = "BCD"
allowed1 = ["BCC", "CDE", "CEA", "FFF"]
result1 = solution.pyramidTransition(bottom1, allowed1)
print(f"Test 1: bottom='{bottom1}', allowed={allowed1}")
print(f"Output: {result1}, Expected: True\n")
